keps/sig-apps/3939-allow-recreation-when-fully-terminated/README.md

```diff
@@ -0,0 +1,696 @@
+
+If a job is stuck in terminating, it could be possible for autoscaling to kick in and give a new node.  
+This is not ideal if you could just wait for the pod to terminating and reuse that node.  
+
+### Goals
+
+- Job controller should allow for flexibility in waiting for pods to be fully terminated before
+  creating new ones
+- Job controller will have a new status field where we include the number of terminating pods.
+
+### Non-Goals
+
+- Other workload APIs are not included in this proposal.
+
+## Proposal
+
+The Job controller gets a list of active pods.  Active pods usually mean pods that have not been registered for deletion.  
+In this KEP, we will consider terminating pods to be separate from active and failed.  
+In the job controller, we should include a field that states the number of terminating pods.  
+
+We propose two new API fields:
+
+1) A field in Spec that allows for opt-in behavior of whether to wait for terminating pods to finish before recreating.
+2) A new field in Status for tracking the number of terminating pods.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a machine learning user, ML frameworks allow scheduling of multiple pods.  
+The Job controller does not typically wait for terminating pods to be marked as failed.  
+Tensorflow and other ML frameworks may have a requirement that they only want Pods to be started once the other pods are fully terminated.
+
+This case was added due to a bug discovered with running IndexedJobs with Tensorflow.  
+See [Jobs create replacement Pods as soon as a Pod is marked for deletion](https://github.com/kubernetes/kubernetes/issues/115844) for more details.
+
+#### Story 2
+
+As a cloud user, users would want to guarantee that the number of pods that are running is exactly the amount that they specify.  
+Terminating pods do not relinguish resources so scarce compute resource are still scheduled to those pods.
+See [Kueue: Account for terminating pods when doing preemption](https://github.com/kubernetes-sigs/kueue/issues/510) for an example of this.
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+One area of contention is how this KEP will work with 3329-retriable-and-non-retriable-failures.  It is important to mention the subtleties here.  
+In 3329, there was a decision to force kubelet to transition pods to failed before marking them as failed.  This is feature toggled guarded by `PodDisruptionCondition`.  
+This means that when this feature is turned on we will only mark pods as failed once they are fully terminated.  This means that the pod is still considered running, as opposed to failed.  
+If `PodDisruptionCondition` is turned off then a pod is marked as failed as soon as it is terminated so it will be counted as failed rather than active.  
```

Comment:
```suggestion
If `PodDisruptionCondition` is turned off then the job controller considers the pod as failed as soon as it is terminating (has a deletion timestamp), because there is no guarantee that the pod will transition to phase=Failed.
```

Reply:
done.